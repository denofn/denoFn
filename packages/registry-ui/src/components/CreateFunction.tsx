import { useAuth0 } from "@auth0/auth0-react";
import { css } from "otion";
import React from "react";
import { useFetch } from "use-http";

import { RegistryJSON } from "../types";
import { API_V1 } from "../utils/prefixes";
import { Button } from "./Button";
import { CenterAlign } from "./CenterAlign";
import { Code } from "./Code";
import { Error } from "./Error";
import { Pancake } from "./Pancake";
import { Registry } from "./Registry";

export type State = {
  scriptName: string;
  index: string;
  registry: RegistryJSON;
};

export type Action = {
  type: "warmupOnStart" | "index" | "whitelist" | "scriptName";
  value: boolean | string | string[];
};

function functionReducer(state: State, action: Action): State {
  switch (action.type) {
    case "scriptName":
      return { ...state, scriptName: action.value as string };
    case "index":
      return { ...state, index: action.value as string };
    case "warmupOnStart":
      return {
        ...state,
        registry: {
          ...state.registry,
          warmupOnStart: action.value as boolean,
        },
      };
    case "whitelist":
      return {
        ...state,
        registry: {
          ...state.registry,
          whitelist: action.value as string[],
        },
      };
    default:
      return state;
  }
}

export function CreateFunction({ reload }: { reload: () => void }) {
  const { user } = useAuth0();
  const [state, dispatch] = React.useReducer(functionReducer, {
    scriptName: "",
    index: `import {
  createHandler,
  createResponse,
} from "https://cdn.jsdelivr.net/gh/jeroenptrs/denofn@1.0.0/packages/micro/functions/mod.ts";

export default createHandler(async (req, res) => {
  return createResponse({
    body: "Hello, World",
  });
});
`,
    registry: {
      whitelist: [],
      warmupOnStart: false,
    },
  });
  const { post, loading, response, error, data } = useFetch(API_V1(`/functions`), {
    headers: {
      "x-denofn-user": user?.sub,
    },
  });

  async function saveChanges() {
    await post(state);
    if (response.ok) reload();
  }

  function handleChange(e: React.ChangeEvent<HTMLInputElement>): void {
    dispatch({ value: e.target.value, type: "scriptName" });
  }

  return (
    <>
      {error && (
        <CenterAlign>
          <Error title="Something went wrong while creating your new function" message={data} />
        </CenterAlign>
      )}
      <Pancake
        padding="1.5rem"
        flex={false}
        className={css({
          minHeight: "calc(75vh - 7.5rem)",
          marginTop: error ? "1.5rem" : "0",
        })}
      >
        <div
          className={css({
            display: "flex",
            alignItems: "center",
            justifyContent: "space-between",
            margin: "0",
            marginBottom: "1.5rem",
          })}
        >
          <input
            className={css({
              width: "100%",
              padding: ".25rem .5rem",
              lineHeight: "1.5rem",
              borderStyle: "none",
              backgroundColor: "transparent",
              appearance: "none",
              fontSize: "1.5rem",
              fontFamily: "'Rubik', sans-serif",
              fontWeight: "bold",
              "::placeholder": {
                fontFamily: "'Karla', sans-serif",
                fontWeight: "initial",
              },
            })}
            value={state.scriptName}
            onChange={handleChange}
            placeholder="Enter your scriptname here"
            autoFocus
          />
          <Button
            disabled={state.scriptName === ""}
            title="Create"
            onClick={saveChanges}
            loading={loading}
            className={css({ marginLeft: "1.5rem" })}
          />
        </div>
        <hr color="#e2e8f0" />
        <Registry registry={state!.registry} dispatch={dispatch} />
        <Code index={state!.index} dispatch={dispatch} />
      </Pancake>
    </>
  );
}
