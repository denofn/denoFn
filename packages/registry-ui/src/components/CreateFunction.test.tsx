import { render, cleanup } from "@testing-library/react";
import React from "react";

import { CreateFunction } from "./CreateFunction";

afterEach(cleanup);

it("renders with the correct styles", () => {
  const { asFragment } = render(<CreateFunction reload={() => {}} />);
  expect(asFragment()).toMatchSnapshot();
});
