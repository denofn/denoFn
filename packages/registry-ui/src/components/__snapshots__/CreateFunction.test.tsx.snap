// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders with the correct styles 1`] = `
<DocumentFragment>
  <div
    class="_4zleql _1qesdsb _1khn195 _sy4y6p _1x4v58x _1m3j4vw _u68uts _1xgdlcu _1i9vogi"
  >
    <div
      class="_zjik7 _1h3rtzg _2o6p8u _ti75j2 _1h026t7"
    >
      <input
        class="_1d3w5wq _1dbsnr2 _1dw7ik1 _ksse9g _1peqg37 _1kzvn2t _1bjds75 _1k2w1pl _1efi8gv _ixc7mc _5rxwa"
        placeholder="Enter your scriptname here"
        value=""
      />
      <button
        class="_1peqg37 _1efi8gv _1k9emg3 _my9kfw _fxc8cu _g3a0px _sy4y6p _r24itg _rkiwex _e0dnmk _1ijvbgw _gxlvbl _ryz482 _yoqe5q _141w5m3 _1y0fyy9 _1g8t4a7 _1n02zsb _x4n4zi"
        disabled=""
      >
        <div
          class="_zjik7 _1h3rtzg"
        >
          Create
        </div>
      </button>
    </div>
    <hr
      color="#e2e8f0"
    />
    <div
      class="_di48ht"
    >
      <div
        class="_1efi8gv _1hl83va"
      >
        URL Whitelist
      </div>
      <div
        class=" css-2b097c-container"
      >
        <div
          class=" css-yk16xz-control"
        >
          <div
            class=" css-g1d714-ValueContainer"
          >
            <div
              class=" css-1wa3eu0-placeholder"
            >
              Enter the domains you want to send requests to here...
            </div>
            <div
              class="css-b8ldur-Input"
            >
              <div
                class=""
                style="display: inline-block;"
              >
                <input
                  aria-autocomplete="list"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  id="react-select-2-input"
                  spellcheck="false"
                  style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
                <div
                  style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                />
              </div>
            </div>
          </div>
          <div
            class=" css-1hb7zxy-IndicatorsContainer"
          />
        </div>
      </div>
      <div
        class="_1fo50kt _zjik7 _1h3rtzg"
      >
        <label
          class="_1gz641 _1efi8gv"
          for="warmupOnStart"
        >
          Post-registration warmup
        </label>
        <input
          id="warmupOnStart"
          type="checkbox"
        />
      </div>
    </div>
    <div
      class="_di48ht"
    >
      <div
        class="_1szb0m3 _1efi8gv"
      >
        index.ts
      </div>
      <div
        class="_1qnppsc"
        style="position: relative; text-align: left; box-sizing: border-box; padding: 0px; overflow: hidden;"
      >
        <textarea
          autocapitalize="off"
          autocomplete="off"
          autocorrect="off"
          class="npm__react-simple-code-editor__textarea"
          data-gramm="false"
          spellcheck="false"
          style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-size: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: absolute; top: 0px; left: 0px; height: 100%; width: 100%; resize: none; overflow: hidden; padding: 4px 4px 4px 4px;"
        >
          import {
  createHandler,
  createResponse,
} from "https://cdn.jsdelivr.net/gh/jeroenptrs/denofn@1.0.0/packages/micro/functions/mod.ts";

export default createHandler(async (req, res) =&gt; {
  return createResponse({
    body: "Hello, World",
  });
});

        </textarea>
        <pre
          aria-hidden="true"
          style="margin: 0px; border: 0px; background: none; box-sizing: inherit; display: inherit; font-family: inherit; font-size: inherit; font-style: inherit; font-variant-ligatures: inherit; font-weight: inherit; letter-spacing: inherit; line-height: inherit; tab-size: inherit; text-indent: inherit; text-rendering: inherit; text-transform: inherit; white-space: pre-wrap; word-break: keep-all; overflow-wrap: break-word; position: relative; pointer-events: none; padding: 4px 4px 4px 4px;"
        >
          <span
            class="token keyword"
          >
            import
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
  createHandler
          <span
            class="token punctuation"
          >
            ,
          </span>
          
  createResponse
          <span
            class="token punctuation"
          >
            ,
          </span>
          

          <span
            class="token punctuation"
          >
            }
          </span>
           
          <span
            class="token keyword"
          >
            from
          </span>
           
          <span
            class="token string"
          >
            "https://cdn.jsdelivr.net/gh/jeroenptrs/denofn@1.0.0/packages/micro/functions/mod.ts"
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
          


          <span
            class="token keyword"
          >
            export
          </span>
           
          <span
            class="token keyword"
          >
            default
          </span>
           
          <span
            class="token function"
          >
            createHandler
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token keyword"
          >
            async
          </span>
           
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token parameter"
          >
            req
            <span
              class="token punctuation"
            >
              ,
            </span>
             res
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
           
          <span
            class="token operator"
          >
            =&gt;
          </span>
           
          <span
            class="token punctuation"
          >
            {
          </span>
          
  
          <span
            class="token keyword"
          >
            return
          </span>
           
          <span
            class="token function"
          >
            createResponse
          </span>
          <span
            class="token punctuation"
          >
            (
          </span>
          <span
            class="token punctuation"
          >
            {
          </span>
          
    body
          <span
            class="token operator"
          >
            :
          </span>
           
          <span
            class="token string"
          >
            "Hello, World"
          </span>
          <span
            class="token punctuation"
          >
            ,
          </span>
          
  
          <span
            class="token punctuation"
          >
            }
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
          

          <span
            class="token punctuation"
          >
            }
          </span>
          <span
            class="token punctuation"
          >
            )
          </span>
          <span
            class="token punctuation"
          >
            ;
          </span>
          

          <br />
        </pre>
        <style
          type="text/css"
        >
          
/**
 * Reset the text fill color so that placeholder is visible
 */
.npm__react-simple-code-editor__textarea:empty {
  -webkit-text-fill-color: inherit !important;
}

/**
 * Hack to apply on some CSS on IE10 and IE11
 */
@media all and (-ms-high-contrast: none), (-ms-high-contrast: active) {
  /**
    * IE doesn't support '-webkit-text-fill-color'
    * So we use 'color: transparent' to make the text transparent on IE
    * Unlike other browsers, it doesn't affect caret color in IE
    */
  .npm__react-simple-code-editor__textarea {
    color: transparent !important;
  }

  .npm__react-simple-code-editor__textarea::selection {
    background-color: #accef7 !important;
    color: transparent !important;
  }
}

        </style>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
